pkg/registry/resolver.go

```diff
@@ -2,7 +2,7 @@ package registry
 
```

Comment:
Why does adding this validation require us to pull all validation into the resolver? I think the general assumption is that config that reaches the resolver is valid, we validate it before merging anything.

Reply:
IIRC when we looked through this, there was technically a gap here - we validate in ci-op after resolving, so if someone checked in a step that did not get used anywhere, it could remain invalid. There didn't seem to be any ideological reason as to why we only validate on one end of things *and* we likely wanted the validation to occur in both cases (however unlikely that edge was where an unused step was checked in).

Reply:
Yeah I am not saying we don't want it, just that we don't need it in the configresolver but a presubmit instead. Right now this requires exporting a lot of code which isn't great

Reply:
Can we assume that everything that reaches the resolver will for sure be valid? I.e. there is no way that it can change before it reaches the resolver? I don't yet know enough, so that's why I'm asking :). When I was playing around with it it appeared that, as @stevekuznetsov said, you could potentially check in an unused step and that step would then not be validated until it was actually used in a job, which didn't seem ideal.

Reply:
> you could potentially check in an unused step and that step would then not be validated until it was actually used in a job, which didn't seem ideal.

That does sound like something we should fix.

The reason I am not much in favor of doing this validation in the configresolver is that the only ones who might ever see an error there is our team, even though we won't own the config that causes the error.

Reply:
I agree, it's very unexpected to see this change in the middle of this PR.

We have `pull-ci-openshift-release-master-ci-operator-registry`, which loads the registry and validates the contents of all files (using `ci-operator-configresolver --validate-only`).  If there is a validation gap, I'd expected it to be fully stated and fixed separately.

Reply:
(that said, this is definitely not the most straightforward part of `ci-tools`, I'd be happy to provide some guidance)